src/config/reader/grpc_fetch.rs

```diff
@@ -0,0 +1,404 @@
+use nom::AsBytes;
+use prost::Message;
+use prost_reflect::prost_types::{FileDescriptorProto, FileDescriptorSet};
+use serde::{Deserialize, Serialize};
+use serde_json::{json, Value};
+
+use crate::grpc::protobuf::{ProtobufOperation, ProtobufSet};
+use crate::grpc::RequestTemplate;
+use crate::mustache::Mustache;
+use crate::runtime::{TargetRuntime, TargetRuntimeContext};
+
+const REFLECTION_PROTO: &str = include_str!(concat!(
+    env!("CARGO_MANIFEST_DIR"),
+    "/proto/reflection.proto"
+));
+
+/// This function is just used for better exception handling
+fn get_protobuf_set() -> Result<ProtobufSet> {
+    let descriptor = protox_parse::parse("reflection", REFLECTION_PROTO)?;
+    let mut descriptor_set = FileDescriptorSet::default();
+    descriptor_set.file.push(descriptor);
+    ProtobufSet::from_proto_file(&descriptor_set)
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Service {
+    name: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct ListServicesResponse {
+    service: Vec<Service>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct FileDescriptorProtoResponse {
+    file_descriptor_proto: Vec<String>,
+}
+
+impl FileDescriptorProtoResponse {
+    fn get(self) -> Result<Vec<u8>> {
+        let file_descriptor_proto = self
+            .file_descriptor_proto
+            .first()
+            .context("Received empty fileDescriptorProto")?;
+
+        Ok(BASE64_STANDARD.decode(file_descriptor_proto)?)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct CustomResponse {
+    list_services_response: Option<ListServicesResponse>,
+    file_descriptor_response: Option<FileDescriptorProtoResponse>,
+}
+
+/// Makes `ListService` request to the grpc reflection server
+pub async fn list_all_files(url: &str, target_runtime: &TargetRuntime) -> Result<Vec<String>> {
+    let protobuf_set = get_protobuf_set()?;
+
+    let grpc_method = "grpc.reflection.v1alpha.ServerReflection.ServerReflectionInfo".try_into()?;
+
+    let reflection_service = protobuf_set.find_service(&grpc_method)?;
+    let operation = reflection_service.find_operation(&grpc_method)?;
+
+    // let mut methods = vec![];
+    let mut url: url::Url = url.parse()?;
+    url.set_path("grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo");
+
+    let req_template = RequestTemplate {
+        url: Mustache::parse(url.as_str())?,
+        headers: vec![(
+            HeaderName::from_static("content-type"),
+            Mustache::parse("application/grpc+proto")?,
+        )],
+        body: Mustache::parse(json!({"list_services": ""}).to_string().as_str()).ok(),
+        operation: operation.clone(),
+        operation_type: Default::default(),
+    };
+
+    let ctx = TargetRuntimeContext {
+        runtime: target_runtime,
+        vars: &Default::default(),
+        headers: Default::default(),
+    };
+
+    let req = req_template.render(&ctx)?.to_request()?;
```

Comment:
that is very similar code across 3 function here and that could be moved to separate function